mixer/adapter/circonus/circonus.go

```diff
@@ -0,0 +1,221 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package circonus
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	cgm "github.com/circonus-labs/circonus-gometrics"
+	"github.com/circonus-labs/circonus-gometrics/checkmgr"
+	"istio.io/istio/mixer/adapter/circonus/config"
+	"istio.io/istio/mixer/pkg/adapter"
+	"istio.io/istio/mixer/template/metric"
+	"log"
+	"net/url"
+	"sync"
+	"time"
+)
+
+type (
+	builder struct {
+		adpCfg      *config.Params
+		metricTypes map[string]*metric.Type
+	}
+
+	handler struct {
+		cm            cgm.CirconusMetrics
+		env           adapter.Env
+		metrics       map[string]config.Params_MetricInfo_Type
+		lastFlushed   time.Time
+		flushLock     sync.Mutex
+		flushInterval time.Duration
+	}
+)
+
+// ensure types implement the requisite interfaces
+var _ metric.HandlerBuilder = &builder{}
+var _ metric.Handler = &handler{}
+
+// bridge stdlog to env.Logger()
+type logToEnvLogger struct {
+	env adapter.Env
+}
+
+// Build constructs a circonus-gometrics instance and sets up the handler
+func (b *builder) Build(ctx context.Context, env adapter.Env) (adapter.Handler, error) {
+
+	// register cgm.Flush() with env.ScheduleWork() to wrap goroutines
+	var cm *cgm.CirconusMetrics
+	ec := make(chan error, 1)
+
+	env.ScheduleWork(
```

Comment:
Oh, is this because NewCirconusMetrics creates goroutines of its own?

I don't believe goroutines 'nest' in that way. If we install a panic handler for one goroutine, if that goroutine spawns another goroutine, that new one will not inherit the panic handler.

So I don't think this goroutine has value.

Reply:
NewCirconusMetric spawns a goroutine for initialization. I wrapped this in ScheduleWork for that reason. If the panic handler doesn't catch the spawned goroutine, agreed that wrapping this doesn't make sense.